8343531: Improve print_location for invalid heap pointers

simonis opened the pull request:
Currently `BlockLocationPrinter<CollectedHeapT>::print_location()` checks for a pointer if it points into the heap and if that's true, it either prints it as an oop if `is_valid_obj()` is true or it tries to find the the start address of an oop for that pointer by calling `CollectedHeapT::heap()->block_start()`.

However, the `block_start()` functionality is not fully implemented for all GCs (e.g. the young generation of `ParallelScavengeHeap`) and for these cases `block_start()` returns NULL. Because of this NULL return value `os::print_location()` will finally qualify the corresponding pointer as pointing "into unknown readable memory" although we already know that it actually points into an invalid heap area.

In such cases, print at least that the pointer is pointing into an unknown part of the heap instead of just saying that it points into unknown memory. 

I've manually tested the new functionality in GDB.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343531](https://bugs.openjdk.org/browse/JDK-8343531): Improve print_location for invalid heap pointers (**New Feature** - P4)


### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21870/head:pull/21870` \
`$ git checkout pull/21870`

Update a local copy of the PR: \
`$ git checkout pull/21870` \
`$ git pull https://git.openjdk.org/jdk.git pull/21870/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21870`

View PR using the GUI difftool: \
`$ git pr show -t 21870`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21870.diff">https://git.openjdk.org/jdk/pull/21870.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21870#issuecomment-2454242811)
</details>

albertnetymk commented:
> However, the block_start() functionality is not fully implemented for all GCs (e.g. the young generation of ParallelScavengeHeap) and for these cases block_start() returns NULL.

Can we implement it properly for all gcs, instead of working around the issue in the caller?

tschatzl commented on src/hotspot/share/gc/shared/locationPrinter.inline.hpp:
```diff
@@ -51,8 +51,10 @@ oop BlockLocationPrinter<CollectedHeapT>::base_oop_or_null(void* addr) {
 
 template <typename CollectedHeapT>
 bool BlockLocationPrinter<CollectedHeapT>::print_location(outputStream* st, void* addr) {
+  bool in_heap = false;
   // Check if addr points into Java heap.
   if (CollectedHeapT::heap()->is_in(addr)) {
+    // base_oop_or_null() might be unimplemented and return NULL for some GCs/generations
```
In such cases where the flag that we later set is dependent on the complete condition, it seems nicer to assign the result of the condition to it right away. That saves the assignment later too, having only a single assignment to it. Ymmv.
```suggestion
  // Check if addr points into Java heap.
  bool in_heap = CollectedHeapT::heap()->is_in(addr);
  if (in_heap) {
    // base_oop_or_null() might be unimplemented and return NULL for some GCs/generations.
```

(And drop the assignment to `in_heap` later).

simonis replied:
Thanks for looking at this PR. Your suggestion sound like a reasonable simplification. I've updated the code accordingly.

simonis commented:
> > However, the block_start() functionality is not fully implemented for all GCs (e.g. the young generation of ParallelScavengeHeap) and for these cases block_start() returns NULL.
> 
> Can we implement it properly for all gcs, instead of working around the issue in the caller?

Everything is possible :) but I think it is not trivial. The problem is that we can crash at any time. In order to implement it reliably, we would have to make the heap walkable but I don't think we want to do this in the crash handler. Any suggestions?

albertnetymk commented:
> but I think it is not trivial.

I was thinking copying the Serial impl into `ParallelScavengeHeap::block_start`; nothing sophisticated.

I suspect the following oddly looking code is used to workaround the unimplemented branch of block_start.

```
    if (DebuggingContext::is_enabled() || VMError::is_error_reported()) {
      return nullptr;
    }
```

shipilev commented on src/hotspot/share/gc/shared/locationPrinter.inline.hpp:
```diff
@@ -83,6 +85,10 @@ bool BlockLocationPrinter<CollectedHeapT>::print_location(outputStream* st, void
 #endif
 
+  if (in_heap) {
+    st->print_cr(PTR_FORMAT " is an unknown heap location", p2i(addr));
+    return true;
```
So why not put this block as `else` branch in `base_oop_or_null` check at L67? This would also remove any ambiguity whether the in-heap pointer would look like a compressed pointer to object, which would be accidentally handled by the block at L64..L86?

simonis replied:
That was actually the first thing I did. But then I thought that (especially with zero-based compressed oops) we might get quite some valid compressed oops pointers unnecessarily printed as "unknown heap location".
On the other hand, I don't think that there's a high probability for a real invalid heap pointer to be classified as compressed oops pointer because the compressed oops detection code uses `is_valid_obj()` anyway.
So this change is conservative in the sense that it doesn't change any behavior except that pointers which have been printed as pointing "into unknown readable memory" can now be detect as "invalid heap pointers".

If you still think we should prioritize the detection steps differently, please let me know.

shipilev replied:
Oh, OK. Compressed pointers make this whole thing a bit messy. I think current code is not handling the case of compressed interior pointers all that well; IDK if we even have those in Hotspot. 

I think there is an ambiguity between compressed pointers and regular pointers at this level, which we cannot reasonably resolve. E.g. if we have zero-based compressed oops with 2-bit shift and 16 GB heap, passing `0x1000000` as the `addr` here cannot distinguish between cases of "regular pointer, points to `0x1000000`" and "compressed pointer, decodes as `0x4000000`". I guess we would like to print both interpretations. But this is way beyond the scope for this PR.

This version would do meanwhile.

simonis replied:
> Oh, OK. Compressed pointers make this whole thing a bit messy. I think current code is not handling the case of compressed interior pointers all that well; IDK if we even have those in Hotspot.
> 

Yes, that's true. I first thought about calling  `BlockLocationPrinter<CollectedHeapT>::print_location()` recursively for the compressed oops case to avoid code duplication and get the same handling for regular and compressed oops, but that would have been a much larger change.
I think we can have compressed oops in registers, e.g. when GC iterates the heap or when compiled code loads a field but the cases are probably more rare than regular oops.

> I think there is an ambiguity between compressed pointers and regular pointers at this level, which we cannot reasonably resolve. E.g. if we have zero-based compressed oops with 2-bit shift and 16 GB heap, passing `0x1000000` as the `addr` here cannot distinguish between cases of "regular pointer, points to `0x1000000`" and "compressed pointer, decodes as `0x4000000`". I guess we would like to print both interpretations. But this is way beyond the scope for this PR.

That's also true, but remember that  `is_valid_obj()` does quite some checks. So in your example, in order to make it really ambiguous, it would require that at address `0x1000000 + 8` as well as at address `0x4000000 + 8` we have properly aligned,  valid (possibly compressed) pointers into MetaSpace pointing to a valid `Klass` object (which is probably not so common for most adresses). 

> This version would do meanwhile.

Thanks for the review.

shipilev reviewed (commented):
I think this is a fine papercut fix, and implementing block information for all GCs could be tackled separately. I do have a question, though:

simonis commented:
> > but I think it is not trivial.
> 
> I was thinking copying the Serial impl into `ParallelScavengeHeap::block_start`; nothing sophisticated.
> 

Unfortunately, the Serial implementation doesn't really work reliably if running with `-XX:+UseTLAB` (which is the default). If called with a pointer which points into unallocated TLAB buffer,  `ContiguousSpace::block_start_const()` will just crash with a SIGSEGV (or a secondary crash during error reporting when called from `VMError`):
```
#0  0x00007ffff57d78ce in oopDesc::size_given_klass (this=0x7ffde5616c70, klass=0x7ffda2000000) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/oops/oop.inline.hpp:196
#1  0x00007ffff57d7756 in oopDesc::size (this=0x7ffde5616c70) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/oops/oop.inline.hpp:153
#2  0x00007ffff689a421 in ContiguousSpace::block_start_const (this=0x7ffff004c880, p=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/space.cpp:565
#3  0x00007ffff689b7ba in Space::block_start (this=0x7ffff004c880, p=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/space.inline.hpp:43
#4  0x00007ffff60f4144 in GenerationBlockStartClosure::do_space (this=0x7ffff530ef30, s=0x7ffff004c880) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/generation.cpp:191
#5  0x00007ffff5e560c5 in DefNewGeneration::space_iterate (this=0x7ffff004b9c0, blk=0x7ffff530ef30, usedOnly=false) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/serial/defNewGeneration.cpp:674
#6  0x00007ffff60f3527 in Generation::block_start (this=0x7ffff004b9c0, p=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/generation.cpp:200
#7  0x00007ffff60e36e9 in GenCollectedHeap::block_start (this=0x7ffff0038450, addr=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/genCollectedHeap.cpp:884
#8  0x00007ffff60e5b97 in BlockLocationPrinter<GenCollectedHeap>::base_oop_or_null (addr=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/locationPrinter.inline.hpp:41
#9  0x00007ffff60e592b in BlockLocationPrinter<GenCollectedHeap>::print_location (st=0x7ffff0000b60, addr=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/locationPrinter.inline.hpp:56
#10 0x00007ffff60e43bd in GenCollectedHeap::print_location (this=0x7ffff0038450, st=0x7ffff0000b60, addr=0x7ffde5616ca0) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/gc/shared/genCollectedHeap.cpp:1046
#11 0x00007ffff66acb22 in os::print_location (st=0x7ffff0000b60, x=140728451820704, verbose=false) at /priv/simonisv/OpenJDK/Git/jdk21u-dev/src/hotspot/share/runtime/os.cpp:1190
```
And that's again because the heap is in general not *walkable* when we call this function. Making it walkable will fill the remaining TLAB spaces with a dummy int array, but without that, we will just trying to interpret random memory (or NULL if running with `-XX:+ZeroTLAB`) as a `Klass` pointer which is seldomly successful :) 

> I suspect the following oddly looking code is used to workaround the unimplemented branch of block_start.
> 
> ```
>     if (DebuggingContext::is_enabled() || VMError::is_error_reported()) {
>       return nullptr;
>     }
> ```

That "oddly looking code" is actually the proof that `block_start()` only gets called from `VMError` or manually, when natively debugging the VM.

albertnetymk commented:
> And that's again because the heap is in general not walkable when we call this function.

It depends on exactly when this function can be called, and with what arg. I wonder whether it can be called with a pointer to a obj that has not been properly initialized (with klass); if so, the heap is almost never walkable, since allocation is not atomic.

> the Serial implementation doesn't really work reliably

I am curious if other GCs' impl work (more) reliably, with regarding to the tlab example.

simonis commented:
@albertnetymk, I'm fine with further improving error reporting by implementing a more sophisticated version of `block_start()` (and I saw you already started a try in #21919. But I still think this patch is a good improvement over the current situation and it will be still valuable for the corner cases which an improved `block_start()` wont be able to handle. So if you don't have any objections, I plan to push this in a day or so.

@tschatzl are you fine with the latest version of this patch (I think I've addressed your suggestions)? If you don't have any additionl objections, I'll push this PR in a day or so.

simonis commented:
Thanks @albertnetymk, @tschatzl and @shipilev for your reviews.
